[Feature] [config] Fix dependency conflict in seatunnel config when running SeatunnelFlink in idea with local mode#1186

Yves-yuan opened the pull request:
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist

  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/incubator-seatunnel/issues).

  - Name the pull request in the form "[SeaTunnel #XXXX] [component] Title of the pull request", where *SeaTunnel #XXXX* should be replaced by the actual issue number.

  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.

-->

## Purpose of this pull request

This pull request fix the #1186 bug.It add a seatunnel-config-shade module and rename package name of code in seatunnel-config to avoid code conflict when running SeatunnelFlink in Intellij Idea with local mode.
<!-- Describe the purpose of this pull request. For example: This pull request adds checkstyle plugin.-->

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in you PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/developement/NewLicenseGuide.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

Yves-yuan commented:
@garyelephant Please review.

asdf2014 commented on seatunnel-config-shade/pom.xml:
```diff
@@ -0,0 +1,116 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please add a license header here

xleoken commented:
hi @Yves-yuan, 

if you want to test some case, you can use `LocalFlinkExample`. it seems not necessary to introduce a new config module.

`seatunnel-example` module will run successfully in local env. 

![image](https://user-images.githubusercontent.com/95013770/152763482-28ce2580-b57a-42f2-abb1-4d416d5e6e67.png)

Yves-yuan commented:
> 
Hello,leo65536,
I have tested LocalFlinkExample just now ,it won't throw the error because the order of jar com.typesafe.config is in the front of seatunnel-config\target\classes in classpath.
When I run SeatunnelFlink directly,the order of jar com.typesafe.config is after seatunnel-config\target\classes,so it will throw the error.
So I think it's more reliable to fix this problem.

Yves-yuan commented:
@CalvinKirs  Hello,code conflict is resolved,can you help rerunning workflow.

CalvinKirs commented:
> @CalvinKirs Hello,code conflict is resolved,can you help rerunning workflow.

Looks has some errors need resolve, you can look at the log.

Yves-yuan commented:
> @CalvinKirs Hello,code conflict is resolved,can you help rerunning workflow.

> Looks has some errors need resolve, you can look at the log.
![image](https://user-images.githubusercontent.com/11640862/152912395-9a876e24-c217-492b-abe9-d8d073a63fdf.png)

The log shows that `Error creating assembly archive bin: /home/runner/work/incubator-seatunnel/incubator-seatunnel/seatunnel-dist/../seatunnel-config/target/seatunnel-config-2.0.5-SNAPSHOT.jar`
`To see the full stack trace of the errors, re-run Maven with the -e switch.`

I have tested mvn -T 2C -B clean verify -Dmaven.test.skip=false -Dgpg.skip=true in my local env,it works well.
Can you rerun maven with -e switch ? @CalvinKirs

Yves-yuan commented:
![image](https://user-images.githubusercontent.com/11640862/152921788-c8781edf-7419-49ef-924e-034e697f45bc.png)
Hello leo65535,
It seems that the workflow was canceled after 30 minutes.I see that you opened the similar issue in https://github.com/apache/incubator-seatunnel/issues/1202#issue-1126677161.So how can i solve this problem? @leo65535

CalvinKirs commented:
> > @CalvinKirs Hello,code conflict is resolved,can you help rerunning workflow.
> 
> > Looks has some errors need resolve, you can look at the log.
> > ![image](https://user-images.githubusercontent.com/11640862/152912395-9a876e24-c217-492b-abe9-d8d073a63fdf.png)
> 
> The log shows that `Error creating assembly archive bin: /home/runner/work/incubator-seatunnel/incubator-seatunnel/seatunnel-dist/../seatunnel-config/target/seatunnel-config-2.0.5-SNAPSHOT.jar` `To see the full stack trace of the errors, re-run Maven with the -e switch.`
> 
> I have tested mvn -T 2C -B clean verify -Dmaven.test.skip=false -Dgpg.skip=true in my local env,it works well. Can you rerun maven with -e switch ? @CalvinKirs

rerun~

xleoken commented:
you can try following statements to trigger the ci again.
```
git commit --allow-empty -m 'trigger rebuild' 
git push
```

Yves-yuan commented:
> you can try following statements to trigger the ci again.
> 
> ```
> git commit --allow-empty -m 'trigger rebuild' 
> git push
> ```

I have ran these commands,but the workflow didn't start running. 
@leo65535

xleoken commented:
hi @Yves-yuan, can you check this page https://github.com/Yves-yuan/incubator-seatunnel/actions

Yves-yuan commented:
> hi @Yves-yuan, can you check this page https://github.com/Yves-yuan/incubator-seatunnel/actions

![image](https://user-images.githubusercontent.com/11640862/152934644-2c0d05a5-50c5-4326-ba77-9606ceebc20b.png)
Ok,I checked this page just now.

Yves-yuan commented:
@asdf2014 
Hi asdf2014,
the checks is passed,can you help reviewing.

Yves-yuan commented:
@CalvinKirs 
Hi CalvinKirs,
the checks of workflow is canceled cause to timeout,can you help rerunning it?

xleoken reviewed (dismissed):
Thanks @Yves-yuan, 💯

CalvinKirs reviewed (changes requested):
You need to refer to this PR to modify the corresponding license, because there are changes to the package name.

CalvinKirs commented:
> You need to refer to this PR to modify the corresponding license, because there are changes to the package name.

https://github.com/apache/incubator-seatunnel/pull/1177/files

Yves-yuan commented:
> > You need to refer to this PR to modify the corresponding license, because there are changes to the package name.
> 
> https://github.com/apache/incubator-seatunnel/pull/1177/files

@CalvinKirs 
Hi CalvinKirs,
I have changed the path of seatunnel-conf classes in LICENSE file.

CalvinKirs commented on LICENSE:
```diff
@@ -209,8 +209,8 @@ The text of each license is the standard Apache 2.0 license.
-seatunnel-config/src/main/java/com/typesafe/config/impl/PathParser.java        from  https://github.com/lightbend/config
-seatunnel-config/src/main/java/com/typesafe/config/ConfigParseOptions.java     from https://github.com/lightbend/config
\ No newline at end of file
+seatunnel-config/seatunnel-config-shade/src/main/java/com/typesafe/config/impl/ConfigNodePath.java    from  https://github.com/lightbend/config
+seatunnel-config/seatunnel-config-shade/src/main/java/com/typesafe/config/impl/ConfigParser.java      from  https://github.com/lightbend/config
+seatunnel-config/seatunnel-config-shade/src/main/java/com/typesafe/config/impl/Path.java              from  https://github.com/lightbend/config
+seatunnel-config/seatunnel-config-shade/src/main/java/com/typesafe/config/impl/PathParser.java        from  https://github.com/lightbend/config
+seatunnel-config/seatunnel-config-shade/src/main/java/com/typesafe/config/ConfigParseOptions.java     from https://github.com/lightbend/config
```
please check file path

Yves-yuan committed:
Fix #1186
Refactor seatunnel-config module to resolve conflict with flink's com.typesafe.config
Rename package name of Seatunnel-config from com.typesafe.config to org.apache.seatunnel.shade.com.typesafe.config

CalvinKirs reviewed (approved):
Well done
Looking forward to your next contribution

Yves-yuan commented:
> Well done Looking forward to your next contribution

Thx,be willing to.

xleoken commented:
Thanks @CalvinKirs
